### ISSUE TITLE
tests using cql_test_env can fail with raft::request_aborted

### ISSUE THREAD

**Issue by denesb** [scylladb/scylladb#21983]
Last seen in https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13902/testReport/junit/boost/database_test/test_database_with_data_in_sstables_is_a_mutation_source_reverse_reader_conversion_cg0/

```
[Exception] - raft::request_aborted: Abort requested while waiting for entry with idx: 172, term: 1; last committed entry: 172, last applied entry: 171
 == [File] - test/lib/mutation_reader_assertions.hh
 == [Line] - 37
```

Any test using cql_test_env can fail with this, apparently there is something wrong with the shutdown sequence in `cql_test_env`.

**Comment by nyh** [comment]
Seen again in https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14096/testReport/junit/boost/database_test/test_database_with_data_in_sstables_is_a_mutation_source_plain_fragments_monotonic_cg1/

**Comment by Annamikhlin** [comment]
seen again in 
https://jenkins.scylladb.com/job/scylla-master/job/next/8622/testReport/junit/boost/database_test/test_database_with_data_in_sstables_is_a_mutation_source_plain_reader_conversion_cg0/

**Comment by mykaul** [comment]
Do we have an unhandled exception during the shutdown sequence?

**Comment by kostja** [comment]
It's a shutdown order bug . Maybe not even related to raft -raft is stopped before some other system that is using it.
We should stop fixing shutdown order bugs @avikivity , it's a waste of time. Drain then _exit. Including in cql_test_env.

All this time we spend on such issues could be directed to work on real modularity.

**Comment by kbr-scylla** [comment]
>  We should stop fixing shutdown order bugs @avikivity , it's a waste of time. Drain then _exit. Including in cql_test_env.

It's gonna turn all "shutdown bugs" into "drain bugs"; other than that, the bugs will stay exactly the same

**Comment by kostja** [comment]
No, they will not be exactly the same, these are different code paths. The whole point is to reduce the amount of code paths we have to support and actually fix the code paths relevant to the user, not to the C++ god.

**Comment by denesb** [comment]
> No, they will not be exactly the same, these are different code paths. The whole point is to reduce the amount of code paths we have to support and actually fix the code paths relevant to the user, not to a C++ god.

I don't see how not deleting any object, then calling `_exit()` after all `stop()/drain()` finished will solve anything. If there is a bug in the drain/stop sequence we will just paper over it by calling `_exit()`, then later we will possibly find corrupted entries in internal tables or possibly corrupted files because we called `_exit()` in the middle of some operation that was mistakenly started even though drain was initiated.

**Comment by kbr-scylla** [comment]
Let's suppose you have a service A which depends on service B and you stop service B during shutdown while A is still running, A tries to access service B and gets an error such as `request_aborted`, tests sees it, we open an issue like this one.

Now take the description above, replace "shutdown" with "drain". Nothing changes

**Comment by kbr-scylla** [comment]
I need full logs to investigate this, in particular I need to see the shutdown logs and see in which place the failure is happening. But no logs were uploaded to the issue. So I can't investigate

We need to wait for new failure to appear and hopefully whoever reports it will upload the logs before I get assigned to the issue 2 weeks later when the logs are gone

**Comment by nyh** [comment]
How much do we currently pay for 2 weeks of log storage? Could we afford 4 weeks?

**Comment by denesb** [comment]
[boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_reverse_read_back_cg0.1.log](https://github.com/user-attachments/files/18426496/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_reverse_read_back_cg0.1.log)

**Comment by kbr-scylla** [comment]
Thanks, will try digging a bit

**Comment by kostja** [comment]
> Let's suppose you have a service A which depends on service B and you stop service B during shutdown while A is still running, A tries to access service B and gets an error such as `request_aborted`, tests sees it, we open an issue like this one.
> 
> Now take the description above, replace "shutdown" with "drain". Nothing changes

Please read storage_service::drain() code and compare it with the code in main().
You can easily see that it's an order of magnitude smaller effort - to run drain. Take query processor, for example, it's not stopped. Take raft. It's not stopped. So of course, if you theorize about abstract services A and B, it's the same, but in reality, it's not the same. With drain() and _exit it's a vastly smaller problem.

**Comment by kbr-scylla** [comment]
@kostja I believe a correct drain procedure for some service that accepts requests works like this:
- stop accepting new requests
- (possibly) cancel ongoing requests
- wait until all ongoing requests finish

at the end of the procedure, the service should not have any ongoing requests and not accept any new requests.

So if you have A->B dependency and you start draining B and then A sends request to B, then A will get an error because B no longer accepts new requests.

BTW storage_service::drain() runs on shutdown.

**Comment by kbr-scylla** [comment]
so I believe that the correct solution is nested lifetimes. If A depends on B then A must be shutdown before B. Or A must be drained before B, doesn't matter. Otherwise you'll get failed requests anyway.

If you have cyclic dependency then it should work like this:
- start B
- start A, establish dependency on B (can be done in single step, the dependency passed as B& in constructor for example)
- then explicitly establish the reverse dependency B->A (separate step, through setting some pointer or whatever)

and then the reverse order of drain or shutdown should look like this:
- first disable the dependency B->A (stop producing B->A requests, wait for ongoing requests to finish)
- then stop A (stop producing A->B requests etc.)
- then stop B

**Comment by kbr-scylla** [comment]
Regarding the current issue

unfortunately those `compaction_manager` messages are the last ones that are printed during successful shutdown... services stopped after that don't print anything in `cql_test_env` (unlike `main`, `cql_test_env` does not use `defer_verbose_shutdown`) so it's impossible to deduce which service produced the error just by looking at these logs

one interesting thing I noticed is that `compaction_manager` shutdown behaved differently on shard 0 than shard 1, but I don't know if it's relevant. (To be precise, `really_do_stop` -- I love the name by the way -- on shard 0 was not called during this time, it happened earlier close to the beginning of shutdown)

**Comment by kbr-scylla** [comment]
1000 runs in debug mode, running 10 instances of the test in parallel:
```bash
parallel --halt now,fail=1 -j 10 'build/debug/test/boost/combined_tests  --run_test=database_test/test_database_with_data_in_sstables_is_a_mutation_source_reverse_read_back_cg0 --report_level=no --catch_system_errors=no --color_output=false -- -c2 -m2G --overprovisioned --unsafe-bypass-fsync 1 --kernel-page-cache 1 --blocked-reactor-notify-ms 2000000 --collectd 0 --max-networking-io-control-blocks=100 \
 > tmp/logs/eraseme_{}.log 2>&1' ::: {1..1000}
```

didn't reproduce it.

Let's see if it continues happening -- the last failure posted by Botond is from 9 days ago and noone referenced this issue after.

If it continues happening, one approach would be to replace `defer` with `defer_verbose_shutdown` in `cql_test_env` which would help pinpointing the service producing that exception

**Comment by Annamikhlin** [comment]
seen again on https://jenkins.scylladb.com/job/scylla-master/job/next/8753/testReport/junit/boost/database_test/test_database_with_data_in_sstables_is_a_mutation_source_reverse_read_back_cg0/

**Comment by denesb** [comment]
Seen again in: https://jenkins.scylladb.com/job/scylla-master/job/next/8753/

**Comment by kbr-scylla** [comment]
We'll have to modify `cql_test_env` shutdown to be more verbose.

**Comment by tchaikov** [comment]
and in [test_database_with_data_in_sstables_is_a_mutation_source_reverse_basic_cg1](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14724/testReport/junit/boost/database_test/test_database_with_data_in_sstables_is_a_mutation_source_reverse_basic_cg1/):

```
[Exception] - raft::request_aborted: Abort requested while waiting for entry with idx: 46, term: 1; last committed entry: 46, last applied entry: 45
 == [File] - test/lib/mutation_source_test.cc
 == [Line] - 392
```

**Comment by tchaikov** [comment]
and in [test_database_with_data_in_sstables_is_a_mutation_source_plain_read_back_cg1](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14654/testReport/junit/boost/database_test/test_database_with_data_in_sstables_is_a_mutation_source_plain_read_back_cg1/)

```
[Exception] - raft::request_aborted: Abort requested while waiting for entry with idx: 246, term: 1; last committed entry: 246, last applied entry: 245
 == [File] - test/lib/mutation_reader_assertions.hh
 == [Line] - 37
```

**Comment by kbr-scylla** [comment]
> If it continues happening, one approach would be to replace defer with defer_verbose_shutdown in cql_test_env which would help pinpointing the service producing that exception

I opened https://github.com/scylladb/scylladb/pull/22566

**Comment by nodep** [comment]
Seen in https://jenkins.scylladb.com/job/scylla-2025.1/job/scylla-ci/6/testReport/junit/boost/database_test/test_database_with_data_in_sstables_is_a_mutation_source_reverse_reader_conversion_cg0/  

```
Stacktrace
[Exception] - raft::request_aborted: Request aborted while performing action on leader, current leader: aec584fb-631e-4b5c-8386-0cf9a526d870, previous leader: unknown
 == [File] - test/lib/mutation_reader_assertions.hh
 == [Line] - 37
Standard Output
[Info] - 1 out of 9 times failed.
 == [File] - test/boost/database_test.cc
 == [Line] - 375

[Info] - passed: debug=2, release=3, dev=3
 == [File] - test/boost/database_test.cc
 == [Line] - 375

[Info] - failed: debug=1
 == [File] - test/boost/database_test.cc
 == [Line] - 375
Standard Error
[Exception] - raft::request_aborted: Request aborted while performing action on leader, current leader: aec584fb-631e-4b5c-8386-0cf9a526d870, previous leader: unknown
 == [File] - test/lib/mutation_reader_assertions.hh
 == [Line] - 37
```

and in https://jenkins.scylladb.com/job/scylla-2025.1/job/scylla-ci/6/testReport/junit/boost/database_test/test_database_with_data_in_sstables_is_a_mutation_source_reverse_read_back_cg1/

```
Stacktrace
[Exception] - raft::request_aborted: Abort requested while waiting for entry with idx: 182, term: 1; last committed entry: 182, last applied entry: 181
 == [File] - test/lib/mutation_reader_assertions.hh
 == [Line] - 37
Standard Output
[Info] - 1 out of 9 times failed.
 == [File] - test/boost/database_test.cc
 == [Line] - 400

[Info] - passed: debug=2, release=3, dev=3
 == [File] - test/boost/database_test.cc
 == [Line] - 400

[Info] - failed: debug=1
 == [File] - test/boost/database_test.cc
 == [Line] - 400
Standard Error
[Exception] - raft::request_aborted: Abort requested while waiting for entry with idx: 182, term: 1; last committed entry: 182, last applied entry: 181
 == [File] - test/lib/mutation_reader_assertions.hh
 == [Line] - 37
```

**Comment by avikivity** [comment]
Seen again. https://github.com/scylladb/scylladb/pull/22594. That pull request contains febd45861e14a205b8187c117090c08d9f173d0c. @emdroid please debug it.

**Comment by emdroid** [comment]
We can see this in the failing logs from #22594:

```
20:51:57  INFO  2025-02-02 17:56:29,216 [shard 0:main] testlog - Shutting down feature service
20:51:57  INFO  2025-02-02 17:56:29,217 [shard 0:main] testlog - Shutting down feature service was successful
20:51:57  INFO  2025-02-02 17:56:29,217 [shard 0:main] testlog - Shutting down configurables
20:51:57  INFO  2025-02-02 17:56:29,217 [shard 0:main] testlog - Shutting down configurables was successful
20:51:57  unknown location(0): fatal error: in "database_test/test_database_with_data_in_sstables_is_a_mutation_source_reverse_reader_conversion_cg0": raft::request_aborted: Abort requested while waiting for entry with idx: 94, term: 1; last committed entry: 94, last applied entry: 93
20:51:57  test/lib/mutation_reader_assertions.hh(37): last checkpoint
```

"configurables" actually seems to be the last service to be shut down (the first one being `defer`-ed) ... so it looks like all the services actually shut down successfully.

The error seems to only be happening after all the cql_test_env services are already down. But it looks like there is still some RAFT mutation that can't go through for some reason (probably because of being aborted), which causes the assertion.

**Comment by emdroid** [comment]
Attaching the logs of the failing tests:

[boost.database_test.log](https://github.com/user-attachments/files/18645698/boost.database_test.log)

**Comment by gleb-cloudius** [comment]
> Looking at the code (and the discussion in [#18902](https://github.com/scylladb/scylladb/issues/18902)) isn't it that such aborts might actually be expected in the case of shutdown? What seems to happen here is that the `server_impl::wait_for_entry()` gets up to the point of trying to apply the committed entry, but gets interrupted in the meantime by the `abort_source` (apparently because of shutting down). But that might be actually expected during the shutdown.
> 

Yes, but the idea is that shutdown order should be such that all raft users abort first and in their abort procedure they will wait for ongoing operation to complete, so when raft is stopped there is no ongoing operations.

**Comment by emdroid** [comment]
So in that case I suppose we'll need to find the place that triggers the about source.

It also seems to not be easily reproducible locally (so far tried 1000 runs in debug without repro), but I'll try to add a sleep to see if it makes it to reproduce.

**Comment by emdroid** [comment]
Couldn't reproduce so far, but found some interesting lines in the logs of affected runs:
```
[2025-02-02T19:51:54.872Z] INFO  2025-02-02 17:54:21,165 [shard 0:goss] database - Dropping ks.cf with auto-snapshot
[2025-02-02T19:51:54.872Z] INFO  2025-02-02 17:54:21,167 [shard 0:goss] database - Truncating ks.cf with auto-snapshot
[2025-02-02T19:51:54.872Z] ERROR 2025-02-02 17:54:21,208 [shard 0:comp] storage_service - Shutting down communications due to I/O errors until operator intervention: Disk error: std::system_error (error system:12, Cannot allocate memory)
[2025-02-02T19:51:54.875Z] INFO  2025-02-02 17:54:21,209 [shard 0:comp] storage_service - Stop transport: starts
[2025-02-02T19:51:54.875Z] INFO  2025-02-02 17:54:21,209 [shard 0:comp] migration_manager - stopping migration service
[2025-02-02T19:51:54.875Z] INFO  2025-02-02 17:54:21,212 [shard 0:main] testlog - Shutting down local client state
```

It looks like the shutdown might have been triggered by an out of memory problem (causing disk operation to fail) in the first place.

Then I can also see an `abort_requested_exception` in the raft topology raft_state_monitor_fiber (that is basically the one running the topology coordinator):
```
[2025-02-02T19:51:54.880Z] INFO  2025-02-02 17:54:21,219 [shard 0:main] testlog - Shutting down auth service
[2025-02-02T19:51:54.880Z] INFO  2025-02-02 17:54:21,220 [shard 0:main] testlog - Shutting down auth service was successful
[2025-02-02T19:51:54.881Z] INFO  2025-02-02 17:54:21,220 [shard 0:main] testlog - Shutting down group 0 usage in storage service
[2025-02-02T19:51:54.881Z] INFO  2025-02-02 17:54:21,220 [shard 0:main] raft_topology - raft_state_monitor_fiber aborted with seastar::abort_requested_exception (abort requested)
[2025-02-02T19:51:54.881Z] INFO  2025-02-02 17:54:21,222 [shard 0:main] raft_topology - cleanup fiber aborted
[2025-02-02T19:51:54.881Z] INFO  2025-02-02 17:54:21,222 [shard 0:main] testlog - Shutting down group 0 usage in storage service was successful
[2025-02-02T19:51:54.882Z] INFO  2025-02-02 17:54:21,222 [shard 0:main] testlog - Shutting down storage service address map subscription
[2025-02-02T19:51:54.882Z] INFO  2025-02-02 17:54:21,222 [shard 0:main] testlog - Shutting down storage service address map subscription was successful
[2025-02-02T19:51:54.882Z] INFO  2025-02-02 17:54:21,222 [shard 0:main] testlog - Shutting down group 0 service
```

The raft_state_monitor_fiber is holding the abort_source that is then passed down to the topology coordinator (and if there is the abort exception it the abort source is then triggered to make the topology coordinator stop). I suspect that this is likely the same abort source that is passed down to the `server_impl::wait_for_entry()`. However this might as well be a red herring, as I can see this exception also in all the successful runs locally.

Nevertheless I don't see the first error ("Disk error: std::system_error (error system:12, Cannot allocate memory") in the successful runs.

As I can see this is usually failing in the debug runs, that are slower and also can utilize more memory. So my hypothesis is that the run goes out of available system memory (as the debug builds use the standard system allocator), perhaps because of running too many tests in parallel. Causing the memory error and eventually triggering the final error (however not yet sure how).

**Comment by emdroid** [comment]
It looks like this might be a duplicate of #21968 - I can see this "Abort requested while waiting for entry with idx" there as well, and all the logs that are available in this ticket seem to also have the "Disk error: std::system_error (error system:12, Cannot allocate memory" that is being reported in that issue.

**Comment by emdroid** [comment]
I can reproduce this error consistently with the following injection:
```diff
diff --git a/sstables/sstables.cc b/sstables/sstables.cc
index 58d2d52beb..f2d5970c19 100644
--- a/sstables/sstables.cc
+++ b/sstables/sstables.cc
@@ -2901,10 +2901,18 @@ future<sstring> make_toc_temporary(sstring sstable_toc_name, storage::sync_dir s
     sstring prefix = sstable_toc_name.substr(0, sstable_toc_name.size() - sstable_version_constants::TOC_SUFFIX.size());
     sstring new_toc_name = prefix + sstable_version_constants::TEMPORARY_TOC_SUFFIX;
 
+    auto sstable_rename_file = [](const sstring& from, const sstring& to) -> future<> {
+        utils::get_local_injector().inject("sstable_write_fail_disk_error", []() {
+            throw std::system_error(std::make_error_code(std::errc::not_enough_memory), "Cannot allocate memory");
+        });
+
+        return rename_file(from, to);
+    };
+
     sstlog.debug("Removing by TOC name: {}", sstable_toc_name);
     if (co_await sstable_io_check(sstable_write_error_handler, file_exists, sstable_toc_name)) {
         // If new_toc_name exists it will be atomically replaced.  See rename(2)
-        co_await sstable_io_check(sstable_write_error_handler, rename_file, sstable_toc_name, new_toc_name);
+        co_await sstable_io_check(sstable_write_error_handler, sstable_rename_file, sstable_toc_name, new_toc_name);
         if (sync) {
             co_await sstable_io_check(sstable_write_error_handler, sync_directory, parent_path(new_toc_name));
         }
diff --git a/test/boost/database_test.cc b/test/boost/database_test.cc
index 647c44da5c..ca3dcff168 100644
--- a/test/boost/database_test.cc
+++ b/test/boost/database_test.cc
@@ -318,6 +318,7 @@ static void test_database(void (*run_tests)(populate_fn_ex, bool), unsigned cgs)
             for (auto&& m : partitions) {
                 apply_mutation(e.db(), uuid, m).get();
             }
+            utils::get_local_injector().enable("sstable_write_fail_disk_error", true);
             cf.flush().get();
             cf.get_row_cache().invalidate(row_cache::external_updater([] {})).get();
             return mutation_source([&] (schema_ptr s,
```

I.e. triggering the sstables write error (as in #21968) leads to the "Abort requested while waiting for entry with idx" error eventually (apparently because of the abort source being triggered). It reproduces locally with about 50% ratio (the test fails every time, but in the other cases the final error is just the storage write error and a general abort exception).
